### DISCUSSION TITLE
Clojupyter 0.4 release

### DISCUSSION THREAD

**Discussion by nighcoder** [clojupyter/clojupyter#151] [General]
I'm opening a conversation thread for the latest changes merged into `develop` branch and the steps forward its release and/or merge into the `master` branch.

At current state, most of the work from last year was merged into `develop`, except for the extra functions to `display` namespace.
Ideally, we will add the `update-display` and `clear-display` messages at some point in the future, but for now I think it's better to focus on fixes to execute handler to close #139.

The latest merge includes a couple of bugfixes that can backported to `master` branch if we intend to continue with the two separate branches in the future.

I would also like to rework the kernel restart to get a much faster restart time. I often find myself restarting the kernel while working with notebooks, so it looks like a pretty useful feature/bug fix.

There's still to some things left to do, like fixing the tests, review the examples and docs and update it if needed, adding an entry to the changelog.

All these changes can be gradually added and published in smaller releases, depending on how we set up our release process.
So far @klausharbo published conda packages and docker images of the latest releases. @klausharbo: Do you plan on publishing conda packages and docker images for future versions? 

As for library writers, they will need the jar on clojars. @krukow, I know you got write access on clojars, so that should be pretty straightforward. Can you or @klausharbo add me to the list of people with write access?

Speaking of libraries, the current monolithic build system forces anyone to download the full kernel for any small functionality accessed.  As @metasoarous said in #129, most developers need only a small part of clojupyter to extend on. It's `PMimeConvertible` in his case, and `CommsAtom` and `PMimeConvertible` for widget models.

I can also see the `nrepl-middleware`, the `display` namespace and the build system in `dev` as parts of clojupyter that can be useful   outside of the project and worth publish it on clojars. The rest are kernel internals and end user helper functions meant to be called from a notebook/console that don't add too much value as a library.

**Comment by krukow** [comment]
Thanks for getting this started @nighcoder (and for your impactful contributions).

> At current state, most of the work from last year was merged into develop, except for the extra functions to display namespace.
Ideally, we will add the update-display and clear-display messages at some point in the future, but for now I think it's better to focus on fixes to execute handler to close #139.

Agreed on priorities and not adding too much additional scope to the current batch of work (and instead focus on getting it shipped).

> I would also like to rework the kernel restart to get a much faster restart time. I often find myself restarting the kernel while working with notebooks, so it looks like a pretty useful feature/bug fix.

Also agreed - faster start time would be a big improvement (especially to the first-run user experience). 

> As for library writers, they will need the jar on clojars. @krukow, I know you got write access on clojars, so that should be pretty straightforward. Can you or @klausharbo add me to the list of people with write access?

I can do this. Can you confirm that this is your identity: https://clojars.org/nighcoder ?

> Speaking of libraries, the current monolithic build system forces anyone to download the full kernel for any small functionality accessed. As @metasoarous said in #129, most developers need only a small part of clojupyter to extend on. It's PMimeConvertible in his case, and CommsAtom and PMimeConvertible for widget models.

Not super important to me, but makes sense.

> I can also see the nrepl-middleware, the display namespace and the build system in dev as parts of clojupyter that can be useful outside of the project and worth publish it on clojars. The rest are kernel internals and end user helper functions meant to be called from a notebook/console that don't add too much value as a library.

That sounds directionally right. 

One thing I would like to look in to after "the big merge" is support for https://github.com/clojure/tools.build and deps https://clojure.org/guides/deps_and_cli (which I prefer for my library)

**Reply by nighcoder** [reply]
> Also agreed - faster start time would be a big improvement (especially to the first-run user experience).

That would be great, but even with the changes to the restart sequence, the first start will still be as slow as currently, unless we find something wrong with the start sequence while working on a fast restart.

I noticed while working with clojupyter that performance varies with use and subsequent calls to a function can yield faster results if the JVM can optimize the code under the hood.

On my machine clojupyter takes at least 6 seconds to print the first log line in `-main` relative to what the front end logs at start time, so no matter what we do on our code, I don't know how to cut that upfront delay. Perhaps cutting more on the deps and optimizing the jar?
Afaik, someone correct me if I'm wrong, the `hello world` clojure program takes 1 to 2 seconds to run, so we'll always pay a hefty upfront cost at startup.
One way of mitigating this, would be to allow multiple instances of clojupyter on the same JVM, so only the first kernel spawned would take a long time to load, but all the other kernels can be started much faster as long as one is already running.


> Can you confirm that this is your identity: https://clojars.org/nighcoder ?

Yes, that's me. I'm signing the artifacts deployed on clojars with the same key I do the commits, if you want to double check 😄. 

> One thing I would like to look in to after "the big merge" is support for https://github.com/clojure/tools.build and deps https://clojure.org/guides/deps_and_cli (which I prefer for my library)

I'm not familiar with `tools.build`, but it would need to build the same way as `lein metajar` currently does, i.e. build a jar, write the class path to `MANIFEST.MF` with the relative paths of our dependencies and copy said deps to `target`.

Speaking of `tools.deps`, I've tried to replace the Travis CI setup with a Github workflow for clojure, but apparently `clojure` (the executable) is not available in the  workflow (unlike `lein`). 

> java.io.IOException: Cannot run program "clojure" (in directory "/home/runner/work/clojupyter/clojupyter"): error=2, No such file or directory

Is there a different workflow for using `tools.deps` instead of `lein`? @krukow, you're certainly more advanced with github than I am, do you have any input on the issue?

**Reply by krukow** [reply]
>> Can you confirm that this is your identity: https://clojars.org/nighcoder ?

> Yes, that's me. I'm signing the artifacts deployed on clojars with the same key I do the commits, if you want to double check 😄.

Added!

**Reply by krukow** [reply]
> Speaking of tools.deps, I've tried to replace the Travis CI setup with a Github workflow for clojure, but apparently clojure (the executable) is not available in the workflow (unlike lein).

>> java.io.IOException: Cannot run program "clojure" (in directory "/home/runner/work/clojupyter/clojupyter"): error=2, No such file or directory

Yup - I've fixed this now: the build uses an action that supports both `lein` and `tools.deps`: https://github.com/clojupyter/clojupyter/runs/4565323700?check_suite_focus=true

**Reply by krukow** [reply]
Workflow: https://github.com/clojupyter/clojupyter/blob/master/.github/workflows/toolsdeps-build.yml

**Reply by nighcoder** [reply]
Thanks a lot! It looks like black magic to me, so I wouldn't have find it easily 😆

**Reply by nighcoder** [reply]
I see that we can add a badge to our Readme. I wonder if it tracks the latest commit automatically.

[![master](https://github.com/clojupyter/clojupyter/actions/workflows/toolsdeps-build.yml/badge.svg)](https://github.com/clojupyter/clojupyter/actions/workflows/toolsdeps-build.yml)

**Reply by krukow** [reply]
Yup! https://github.com/clojupyter/clojupyter/blob/master/README.md (tracks last build)
